### PR TITLE
fix(KONFLUX-13346): Make PipelineRun filtering configurable

### PR DESCRIPTION
**Pull request by hmariset**
Add NOTIFICATION_FILTER_LABELS and NOTIFICATION_FILTER_ANNOTATIONS environment variables to allow processing both PipelinesAsCode and ART-initiated builds. ART builds were previously filtered out due to missing event-type=push label. Configure both filters to process mta-*, mtc-*, oadp-*, logging-* PipelineRuns.

Fixes: KONFLUX-13346

Co-Authored-By: Claude Code

### CODE REVIEW

**Comment by hmariset** [comment]
> ## Review: #724
> **Head SHA:** [619e55a](https://github.com/konflux-ci/notification-service/commit/619e55a4e130fb544cd7e09374e2190d686223ad) **Timestamp:** 2026-05-04T00:00:00Z **Outcome:** approve
> 
> ### Summary
> This PR makes PipelineRun filtering configurable via two new environment variables (`NOTIFICATION_FILTER_LABELS` and `NOTIFICATION_FILTER_ANNOTATIONS`), allowing the controller to process ART-initiated builds in addition to PipelinesAsCode push builds. The implementation is clean, backward-compatible, and well-tested. `IsPushPipelineRun` delegates to the new `ShouldProcessPipelineRun` function, preserving all existing call sites in predicates. The OR semantics between label and annotation filters are correct. Tests cover default behavior, individual filter types, combined filters, and edge cases including malformed input. No security, injection, or correctness concerns were found.
> 
> ### Findings
> #### Medium
> * **[style/conventions]** `internal/controller/pipelinerun_helper.go:151` — `os.Getenv` is called on every invocation of `ShouldProcessPipelineRun`, which runs on every reconciliation event. While functionally correct (env vars are stable during pod lifetime), parsing the filter configuration once at controller startup and injecting it as a dependency would be more idiomatic for a controller-runtime reconciler and avoids repeated string splitting.
>   Remediation: Consider parsing the filter config in `main.go` or the reconciler constructor and passing the parsed filters to `ShouldProcessPipelineRun` as parameters or via a struct field.
> 
> #### Info
> * **[style/conventions]** `internal/controller/predicates.go` — Predicate function names (e.g., `PushPipelineRunCreatedPredicate`) still reference "Push" even though the filter now accepts non-push PipelineRuns. This is pre-existing naming, not introduced by this PR, but may be worth a follow-up rename for clarity.
> 
> ### Footer
> Outcome: approve This review applies to SHA `619e55a4e130fb544cd7e09374e2190d686223ad`. Any push to the PR head clears this review and requires a new evaluation.
> 
> Previous run

tried using sync.Once but it's adding complexity so going with os.Getenv

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -39,16 +39,21 @@ spec:
...
         - --health-probe-bind-address=:8081
-        - --metrics-secure=true         
+        - --metrics-secure=true
         image: controller:latest
         name: manager
+        env:
+        - name: NOTIFICATION_FILTER_LABELS
+          value: "pipelinesascode.tekton.dev/event-type=push"
+        - name: NOTIFICATION_FILTER_ANNOTATIONS
+          value: "art-jenkins-job-url"
```
Comment (amisstea) [inline config/manager/manager.yaml]: We shouldn't include ART specific values in this controller. This env var should be blank by default. We'll define it with our downstream deployment later.

Please also change all "art" values to something generic throughout these changes.

File: README.md [hunk 2]
```diff
@@ -148,6 +148,19 @@ be sent to and the `region` of the AWS account.
...
+
+By default, the controller only processes PipelineRuns with the label `pipelinesascode.tekton.dev/event-type=push`. 
+This can be customized using the following optional environment variables:
+
+| Name | Description | Example |
+| -- | -- | -- |
+| NOTIFICATION_FILTER_LABELS | Comma-separated list of label key=value pairs to filter PipelineRuns | `pipelinesascode.tekton.dev/event-type=push,pipelinesascode.tekton.dev/event-type=manual` |
+| NOTIFICATION_FILTER_ANNOTATIONS | Comma-separated list of annotation keys to filter PipelineRuns by presence | `art-jenkins-job-url,custom-annotation` |
+
+**Note:** If both variables are unset, the controller defaults to filtering by `pipelinesascode.tekton.dev/event-type=push`.
```
Comment (amisstea) [inline README.md]:
```suggestion
**Note:** If both variables are unset, the controller defaults to filtering by the label key pair `pipelinesascode.tekton.dev/event-type=push`. 
```

File: config/manager/manager.yaml (hunk 1 above)
Comment (avi-biton) [reply config/manager/manager.yaml]: We should also remove all `art` values from the README

**Comment by avi-biton** [comment]
The unit tests for `ShouldProcessPipelineRun` are thorough and cover all the important cases (default fallback, label-only, annotation-only, combined OR logic, edge cases like malformed input and whitespace). 
It is a good coverage.

One gap worth addressing: 
the integration tests in `notificationservice_controller_test.go` only cover the default fallback behavior (no env vars set, push-label PipelineRun). 
There are no integration-level tests that set `NOTIFICATION_FILTER_LABELS` or `NOTIFICATION_FILTER_ANNOTATIONS` and verify the full reconcile flow end-to-end.

It would be valuable to add these tests

File: README.md (hunk 2 above)
Comment (hmariset) [reply README.md]: updated

File: internal/controller/notificationservice_controller_test.go [hunk 3]
```diff
@@ -417,6 +417,67 @@ var _ = Describe("NotificationService Controller", func() {
...
+		annotatedPipelineRunLookupKey := types.NamespacedName{Name: annotatedPipelineRunName, Namespace: namespace}
+
+		Context("when a PipelineRun matches the annotation filter", func() {
+			It("should reconcile and process the PipelineRun", func() {
+				GinkgoT().Setenv("NOTIFICATION_FILTER_ANNOTATIONS", "custom-build-trigger")
+
+				err := nsr.SetupWithManager(k8sManager)
+				Expect(err).ToNot(HaveOccurred())
+
+				notificationsCounter = testutil.ToFloat64(notifications)
```
Comment (avi-biton) [inline internal/controller/notificationservice_controller_test.go]: We can remove it, it is not been tested in this scope

**Comment by avi-biton** [comment]
Only one minor comment, otherwise LGTM

File: config/manager/manager.yaml (hunk 1 above)
Comment (hmariset) [reply config/manager/manager.yaml]: removed

File: internal/controller/notificationservice_controller_test.go (hunk 3 above)
Comment (hmariset) [reply internal/controller/notificationservice_controller_test.go]: removed

**Review by avi-biton — APPROVED** [review]
